mount/mount_id_mapped.go

```diff
@@ -0,0 +1,129 @@
+	}
+	cID, err := strconv.ParseUint(parts[0], 0, 32)
+	if err != nil {
+		return []syscall.SysProcIDMap{}, fmt.Errorf("invalid container id for user namespace remapping, %w", err)
+	}
+	hID, err := strconv.ParseUint(parts[1], 0, 32)
+	if err != nil {
+		return []syscall.SysProcIDMap{}, fmt.Errorf("invalid host id for user namespace remapping, %w", err)
+	}
+	size, err := strconv.ParseUint(parts[2], 0, 32)
+	if err != nil {
+		return []syscall.SysProcIDMap{}, fmt.Errorf("invalid size for user namespace remapping, %w", err)
+	}
+
+	return []syscall.SysProcIDMap{
+		{
+			ContainerID: int(cID),
+			HostID:      int(hID),
+			Size:        int(size),
+		},
+	}, nil
+}
+
+func mountIDMapped(target string, pid int) (err error) {
+	var (
+		path       string
+		attr       unix.MountAttr
+		userNsFile *os.File
+		targetDir  *os.File
+	)
+
+	path = fmt.Sprintf("/proc/%d/ns/user", pid)
+	if userNsFile, err = os.Open(path); err != nil {
+		return fmt.Errorf("Unable to get user ns file descriptor for - %s, %w", path, err)
+	}
+
+	attr.Attr_set = unix.MOUNT_ATTR_IDMAP
+	attr.Attr_clr = 0
+	attr.Propagation = 0
+	attr.Userns_fd = uint64(userNsFile.Fd())
+
+	defer userNsFile.Close()
+	if targetDir, err = os.Open(target); err != nil {
+		return fmt.Errorf("Unable to get mount point target file descriptor - %s, %w", target, err)
+	}
+
+	defer targetDir.Close()
+	return unix.MountSetattr(int(targetDir.Fd()), "", unix.AT_EMPTY_PATH|unix.AT_RECURSIVE, &attr)
+}
+
+// MapMount applies GID/UID shift according to gidmap/uidmap for target path
+func MapMount(uidmap string, gidmap string, target string) (err error) {
+	const (
+		userNsHelperBinary = "/bin/true"
+	)
+	// TODO: Avoid dependency on /bin/true or do in a completely different way
+	// Currently there is no way to pass idmapping directly to mount_setattr,
+	// this is not very convenient from the container runtime point of view.
+	// The id remapping procedure should be done in containerd, due to we have
+	// old approach that use recursive chown.
+	// Maybe it is necessary to think about moving of container rootfs ownership
+	// adjustment to runc due to runc has information about container user namespace.
+	// But personally I think that it would be better to add possibility to call
+	// mount_setattr with explicit id mappings and leave container runtime components
+	// responsibilities unchanged.
+	cmd := exec.Command(userNsHelperBinary)
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Cloneflags: syscall.CLONE_NEWUSER,
+	}
+
+	if cmd.SysProcAttr.UidMappings, err = parseIDMapping(uidmap); err != nil {
+		return err
+	}
+	if cmd.SysProcAttr.GidMappings, err = parseIDMapping(gidmap); err != nil {
+		return err
+	}
+
+	if err = cmd.Start(); err != nil {
+		return fmt.Errorf("Failed to run the %s helper binary, %w", userNsHelperBinary, err)
+	}
+
+	defer func() {
+		if err = cmd.Wait(); err != nil {
```

Comment:
if waitErr := cmd.Wait(); waitErr != nil {
                        err = errors.Wrapf(waitErr, "Failed to run the %s helper binary", userNsHelperBinary)
                }
```suggestion
		if err = cmd.Wait(); err != nil {
		       err = fmt.Errorf("Failed to run the %s helper binary, %w", userNsHelperBinary, err)
		}
```

Reply:
otherwise err instance as return argument would be overwritten at any case at return from function MapMount, and we won't know any error e.g. from mount_setattr.

Reply:
Yes, sure thanks!